src/components/panel/panel.js

```diff
@@ -1636,6 +1685,91 @@ MdPanelRef.prototype._animateClose = function() {
   });
 };
 
+/**
+ * Registers a listener with the panel. The callback should return a promise,
+ * which will allow the action to continue when it gets resolved, or will
+ * prevent an action if it is rejected.
+ * @param {string} type Type of listener.
+ * @param {!angular.$q.Promise<!any>} callback Callback to be registered.
+ * @returns {!MdPanelRef}
+ */
+MdPanelRef.prototype.registerListener = function(type, callback) {
```

Comment:
@crisbeto We had debated following the Event based model, but promises worked differently so we went with this API. (We meaning my team since I made a similar API internally and they had to debate/approve.)